src/libraries/System.Private.CoreLib/src/System/Guid.cs

```diff
@@ -802,58 +875,72 @@ private static bool TryParseExactX(ReadOnlySpan<char> guidString, ref GuidResult
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static byte DecodeByte(char ch1, char ch2, ref int invalidIfNegative)
+        private static byte DecodeByte<TChar>(TChar ch1, TChar ch2, ref int invalidIfNegative) where TChar : unmanaged, IUtfChar<TChar>
         {
             ReadOnlySpan<byte> lookup = HexConverter.CharToHexLookup;
             Debug.Assert(lookup.Length == 256);
+            Debug.Assert(ch1 is byte or char);
```

Comment:
```suggestion
            Debug.Assert(typeof(TChar) == typeof(char) || typeof(TChar) == typeof(char));
```
Assert on type instead of value.

Reply:
Is there a specific reason for that besides the case where `TChar` is `object` and `ch1` is unboxed?

Reply:
Mostly consistency with number formatting.